Configurable masking and hiding for the env endpoint

timyates opened the pull request:
- Changes the env endpoint to be disabled by default (requires user to enable in config)
- If enabled, all env endpoint values are masked with *****
- Allow user to add a bean implementing EnvironmentEndpointFilter which:
  - allows legacy operation to be allowed so it behaves as before
  - can be set to mask or reveal all
  - allows patterns for masking to be defined if reveal all is set
  - allows patterns for plain-text values if mask all is set

The call to the filter also contains the Principal if there is one, so the user can choose
to alter visibility of items based on the currently logged in user

timyates committed:
Configurable masking and hiding for the env endpoint

- Changes the env endpoint to be disabled by default (requires user to enable in config)
- If enabled, all env endpoint values are masked with *****
- Allow user to add a bean implementing EnvironmentEndpointFilter which:
  - allows legacy operation to be allowed so it behaves as before
  - can be set to mask or reveal all
  - allows patterns for masking to be defined if reveal all is set
  - allows patterns for plain-text values if mask all is set

The call to the filter also contains the Principal if there is one, so the user can choose
to alter visibility of items based on the currently logged in user

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/micronaut-projects/micronaut-core?pullRequest=6711) <br/>All committers have signed the CLA.

timyates commented on management/src/main/java/io/micronaut/management/endpoint/env/EnvironmentEndpoint.java:
```diff
@@ -17,104 +17,116 @@
 
```
This is a breaking change as it is a modification to the public API.  Discussed with @jameskleeh and we decided we should be ok, as we couldn't see why someone would be creating one of these themselves, nor extending it for their own use.

timyates commented on management/src/main/java/io/micronaut/management/endpoint/env/EnvironmentFilterSpecification.java:
```diff
@@ -0,0 +1,121 @@
+     *
+     * @return itself for chaining calls.
+     */
+    public EnvironmentFilterSpecification legacyMasking() {
+        allMasked = false;
+        maskedPatterns.addAll(
+                Arrays.stream(PROPERTY_NAMES_TO_MASK)
+                        .map(s -> Pattern.compile(".*" + s + ".*", Pattern.CASE_INSENSITIVE))
+                        .collect(Collectors.toList())
+        );
+        return this;
+    }
```
Are we ok compiling these patterns each request?  We could compile them at startup, or could use something like an AtomicReference compile them on demand and cache them

graemerocher replied:
Makes sense to compile then on startup

timyates replied:
Done.  I used a Stream to compile them all.  Just a plain repetitive array might be quicker, but with only 6 entries, I'm not sure by how much...

graemerocher commented on management/src/main/java/io/micronaut/management/endpoint/env/EnvironmentEndpoint.java:
```diff
@@ -17,104 +17,116 @@
+     * @param environmentFilter The registered {@link EnvironmentEndpointFilter} bean if one is registered
      */
-    public EnvironmentEndpoint(Environment environment) {
+    public EnvironmentEndpoint(Environment environment, @Nullable EnvironmentEndpointFilter environmentFilter) {
```
Need to add a new constructor and not modify this one to avoid breaking binary compatiblity

timyates replied:
I tried with both, and the `EnvironmentEndpointSpec` tests failed (I guess because the one arg constructor got used even though an `EnvironmentEndpointFilter` was defined as a bean)

Is there a way to get it to prefer the two-arg constructor?  Would `@Inject` help?  I'll see

timyates replied:
I think `@Inject` does it! 😎  [`7bc7553` (#6711)](https://github.com/micronaut-projects/micronaut-core/pull/6711/commits/7bc7553cef4cd65382f892ae950339dcd7042495)

graemerocher commented on management/src/main/java/io/micronaut/management/endpoint/env/EnvironmentEndpointFilter.java:
```diff
@@ -0,0 +1,33 @@
+     *
+     * @param specification a specification of which properties are masked or hidden from the endpoint.
+     */
+    void specifyFiltering(EnvironmentFilterSpecification specification);
```
Add `@NonNull` to argument

timyates replied:
Done 👍

graemerocher commented on management/src/main/java/io/micronaut/management/endpoint/env/EnvironmentFilterSpecification.java:
```diff
@@ -0,0 +1,121 @@
+     *
+     * @return itself for chaining calls.
+     */
+    public EnvironmentFilterSpecification maskAll() {
```
Add `@NonNull` to return type

timyates replied:
👍

graemerocher commented on management/src/main/java/io/micronaut/management/endpoint/env/EnvironmentFilterSpecification.java:
```diff
@@ -0,0 +1,121 @@
+     *
+     * @return itself for chaining calls.
+     */
+    public EnvironmentFilterSpecification maskNone() {
```
Add `@NonNull` to return type

timyates replied:
👍

graemerocher commented on management/src/main/java/io/micronaut/management/endpoint/env/EnvironmentFilterSpecification.java:
```diff
@@ -0,0 +1,121 @@
+     * @param propertySourceNameMask one or more Patterns to match against property names for masking.
+     * @return itself for chaining calls.
+     */
+    public EnvironmentFilterSpecification maskPatterns(Pattern... propertySourceNameMask) {
```
Add `@NonNull` to return type

timyates replied:
👍

graemerocher commented on management/src/main/java/io/micronaut/management/endpoint/env/EnvironmentFilterSpecification.java:
```diff
@@ -0,0 +1,121 @@
+     *
+     * @return itself for chaining calls.
+     */
+    public EnvironmentFilterSpecification legacyMasking() {
```
Add `@NonNull` to return type

timyates replied:
👍

graemerocher commented on management/src/main/java/io/micronaut/management/endpoint/env/EnvironmentFilterSpecification.java:
```diff
@@ -0,0 +1,121 @@
+ * @author Tim Yates
+ * @since 3.3.0
+ */
+public class EnvironmentFilterSpecification {
```
Looks like this can be made final

timyates replied:
👍

graemerocher commented on src/main/docs/guide/management/providedEndpoints/environmentEndpoint.adoc:
```diff
@@ -1,20 +1,66 @@
+}
+----
+
+The link:{api}/io/micronaut/management/endpoint/env/EnvironmentFilterSpecification.html[EnvironmentFilterSpecification] class
```
You can use `api:management.endpoint.env.EnvironmentFilterSpecification[]`

timyates replied:
Nice!

graemerocher commented on src/main/docs/guide/management/providedEndpoints/environmentEndpoint.adoc:
```diff
@@ -1,20 +1,66 @@
+also contains the current `Principal` if one exists, so this can be used to further limit access based on the
+currently logged-in user.
+
+Sensible defaults can be applied by calling the link:{api}/io/micronaut/management/endpoint/env/EnvironmentFilterSpecification.html#legacyMasking--[`legacyMasking()`] method.
```
Same here use `api:` macro

timyates replied:
👍

graemerocher commented on src/main/docs/guide/management/providedEndpoints/environmentEndpoint.adoc:
```diff
@@ -1,20 +1,66 @@
 ----
 
+By default the endpoint will mask all values.
+To customize this masking you need to supply a Bean that implements link:{api}/io/micronaut/management/endpoint/env/EnvironmentEndpointFilter.html[EnvironmentEndpointFilter].
```
You can use `api:management.endpoint.env.EnvironmentEndpointFilter[]`

timyates replied:
👍

ilopmar commented on management/src/main/java/io/micronaut/management/endpoint/env/EnvironmentEndpointFilter.java:
```diff
@@ -0,0 +1,33 @@
+ * displayed in the {@link EnvironmentEndpoint}.
+ *
+ * @author Tim Yates
+ * @since 3.2.5
```
Since 3.3.0

timyates replied:
Nice catch!  Teach me to guess the version before asking 😉

timyates committed:
Address first round of feedback

timyates committed:
Fix tests

timyates committed:
Fix since tag

timyates committed:
Add single arg constructor

For binary compatibility.  And annotate new constructor with Inject
so it is used

timyates commented:
@graemerocher Hope I addressed all your feedback 🤞

graemerocher commented:
@timyates from my side yeah, @jameskleeh needs to review next

jameskleeh commented on management/src/main/java/io/micronaut/management/endpoint/env/EnvironmentFilterSpecification.java:
```diff
@@ -0,0 +1,117 @@
+     * @return itself for chaining calls.
+     */
+    public @NotNull EnvironmentFilterSpecification maskPatterns(Pattern... propertySourceNameMask) {
+        maskedPatterns.addAll(Arrays.asList(propertySourceNameMask));
```
This would have no effect without `allMasked = false;` so that should probably be set here

timyates replied:
If `allMasked = true`, then this sets the patterns that are allowed to be in the clear

https://github.com/micronaut-projects/micronaut-core/pull/6711/files#diff-4e7e5d0b9696543f54847a8d042130fc33e81143e23d6c414504d106ec0fa4c7R15

`maskPatterns` is probably a rubbish name...  But `toggleMask` was wrong too, as it's not a toggle, it's additive

jameskleeh replied:
I think `maskPatterns` should define the patterns that are masked, thus only making it useful if not all keys are masked

timyates replied:
With a separate list for unmasked patterns, that only come into play if allMasked is true?

Or we could name it "excludedPatterns" which are excluded from whichever mode you're in... And change the method to exclude() ?

jameskleeh replied:
That sounds OK to me. Need to support static keys as well

jameskleeh replied:
Also the ability to supply predicates. Everything could be converted to predicates

timyates replied:
Switch to using `Predicate<String>` for all `exclusions`

To avoid `unchecked generic array creation for varargs parameter` warnings in user code (due to the `Predicate<String>` varargs), introduced `EnvironmentFilterNamePredicate` as a typealias

Also added a factory method to generate a predicate from a pattern and optional flags

jameskleeh commented on management/src/main/java/io/micronaut/management/endpoint/env/EnvironmentFilterSpecification.java:
```diff
@@ -0,0 +1,117 @@
+     */
+    public @NotNull EnvironmentFilterSpecification legacyMasking() {
+        allMasked = false;
+        maskedPatterns.addAll(LEGACY_MASK_PATTERNS);
```
This should just call `maskPatterns`

timyates replied:
Renamed to `exclusions` as discussed 👍

jameskleeh reviewed (changes requested):
I think this is missing:

- the ability to mask all properties except a set a patterns
- the ability to mask all properties except a set of static keys
- the ability to mask a set of static keys

timyates committed:
Switch to Predicate<String> in place of Patterns

timyates committed:
Checkstyle fixes

timyates committed:
Move env endpoint into disabled section of doc

graemerocher commented on management/src/main/java/io/micronaut/management/endpoint/env/EnvironmentEndpoint.java:
```diff
@@ -17,104 +17,125 @@
      */
     @Read
-    public Map<String, Object> getProperties(@Selector String propertySourceName) {
+    public Map<String, Object> getProperties(@Selector String propertySourceName, @Nullable Principal principal) {
```
This is breaking binary compatibility. Add another deprecated method with the original signature.

timyates replied:
Done

graemerocher commented on management/src/main/java/io/micronaut/management/endpoint/env/EnvironmentEndpoint.java:
```diff
@@ -17,104 +17,125 @@
      */
     @Read
-    public Map<String, Object> getEnvironmentInfo() {
+    public Map<String, Object> getEnvironmentInfo(@Nullable Principal principal) {
```
This is breaking binary compatibility. Add another deprecated method with the original signature.

timyates replied:
Done

jameskleeh replied:
I would rather we just not handle the principal rather than adding a new method. Even adding the method is a breaking change because the original method and any extensions of it would never be called

graemerocher commented on management/src/main/java/io/micronaut/management/endpoint/env/EnvironmentFilterSpecification.java:
```diff
@@ -0,0 +1,131 @@
+        return s -> Pattern.compile(pattern, flags).matcher(s).matches();
+    }
+
+    public interface EnvironmentFilterNamePredicate extends Predicate<String> {
```
This appears to be now part of the public API, therefore needs javadoc

timyates replied:
Done

graemerocher commented on management/src/main/java/io/micronaut/management/endpoint/env/EnvironmentFilterSpecification.java:
```diff
@@ -0,0 +1,131 @@
+        return allMasked ? FilterResult.MASK : FilterResult.PLAIN;
+    }
+
+    public static EnvironmentFilterNamePredicate regularExpressionPredicate(String pattern) {
+        return regularExpressionPredicate(pattern, 0);
+    }
+
+    public static EnvironmentFilterNamePredicate regularExpressionPredicate(String pattern, int flags) {
+        return s -> Pattern.compile(pattern, flags).matcher(s).matches();
+    }
```
These appear now to be part of the public API? if it is the case they need javadoc

timyates replied:
Done.  Would you say they were necessary?  I added them to make it simpler to create the Predicates, but can see that they may add noise 🤔

timyates committed:
Add deprecated methods and JavaDoc

timyates commented:
Also, having the `EnvironmentFilterNamePredicate` type-alias as an inner-class makes the javadoc quite long 😐 Should it be made top-level?  Or does someone know of a better way around the `unchecked generic array creation for varargs parameter` warning?

timyates commented:
After a sync up with James, we have a better API... Incoming asap

timyates committed:
Remove principal and simplify API

jameskleeh committed:
Cleanup

jameskleeh committed:
Merge branch '3.3.x' into timyates/environment-endpoint-filtering

jameskleeh commented:
@graemerocher Can you please give this a final look?
